### DISCUSSION TITLE
A strange question about fluid volume

### DISCUSSION THREAD

**Discussion by KimYoungGR** [InteractiveComputerGraphics/SPlisHSPlasH#227] [Q&A]
I built a scence with a box and an emitter. Then I injected 864 and 1728 paricles in the same box separately.
Particle radius is 0.07 and box is 3×3×2.
The strange thing is the depth of 1728 paricles is more bigger than 2*depth of 864 paricles.
Is this a problem with kernel or something?
![image](https://user-images.githubusercontent.com/85207792/213404341-a2c63ac5-a854-44ac-a2ac-1648ef15efa8.png)

**Comment by janbender** [comment]
Can you check if the boundary handling is working properly? If you use:

"boundaryHandlingMethod": 2

Make sure that you set:

"mapInvert": true, 

if the particles should stay in the box. Otherwise the boundary handling will push them out.

**Comment by KimYoungGR** [comment]
Hi Mr bender!
My code is following. I use "boundaryHandlingMethod": 2 and "mapInvert": true.
Doesn't look like it's causing the problem.

"Configuration"
{
"cameraPosition": [0,2,5],
"cameraLookat": [0,0,0],
"timeStepSize": 0.001,
"numberOfStepsPerRenderUpdate":2,
"particleRadius": 0.07,
"density0": 1000,
"simulationMethod": 4,
'gravitation': [0,-9.81,0],
"cfIMethod": 2,
"cflFactor": 1,
"cflMaxTimeStepSize": 0.005,
"maxIterations": 100,
"maxError": 0.05,
"maxIterationsV": 100,
"maxErrorv": 0.1,
"stiffness": 50000,
"exponent": 7,
"velocityUpdateMethod": 0,
"enableDivergenceSolver": true,
"boundaryHandlingMethod": 2
},
"Materials": [
"id":"Fluid"
"viscosityMethod":1,
"viscosity": 0.01,
"dragMethod": 0,
"colorMapType": 1,
"maxEmitterParticles":550000
}
[
"RigidBodies": [
{
"geometryFile": ".. /models/UnitBox. obj",
"translation": [0,1,0],
"rotationAxis": [1, 0, 0],
"rotationAngle"': 0,
"'scale": [3.0, 3.0, 2.0],
"color": [0.1, 0.4, 0.6, 1.0],
"isDynamic": false,
"isWall": true,
'"mapInvert: true,
"mapThickness": 0.0,
"mapResolution": [30,25,20]
}
],
"Emitters":[
{
"width": 8,
"height": 3,
"translation': [-1.2,1,0.0],
"rotationAxis": [1, 0, 0],
"rotationAngle": 0,
"velocity": 5,
"type": 0,
"emitEndTime": 1
}
]
}

**Comment by janbender** [comment]
Sorry, I cannot test this since the file seems not to be valid.

**Reply by KimYoungGR** [reply]
Sorry I edited it again. Other settings are default.
I also have anther question: In SPlisHSPlasH, does a single particle volunm equal to a sphere that Radius is particleRadius?
 {
	"Configuration": 
	{
		"cameraPosition": [0,2,5],
		"cameraLookat": [0,0,0],
		"timeStepSize": 0.001,
		"numberOfStepsPerRenderUpdate": 2,
		"particleRadius": 0.07, 
		"density0": 1000, 
		"simulationMethod": 4,
		"gravitation": [0,-9.81,0], 
		"cflMethod": 2, 
		"cflFactor": 0.5,
		"cflMaxTimeStepSize": 0.005,
		"maxIterations": 100,
		"maxError": 0.05,
		"maxIterationsV": 100,
		"maxErrorV": 0.1,		
		"stiffness": 50000,
		"exponent": 7,
		"velocityUpdateMethod": 0,
		"enableDivergenceSolver": true,
		"boundaryHandlingMethod": 2
	},
	"Materials": [
		{
			"id": "Fluid",
			"maxEmitterParticles": 8000,
			"viscosityMethod": 1,
			"viscosity": 0.01,
			"dragMethod": 0, 
			"colorMapType": 1
		}
	],
	"RigidBodies": [
		{
			"geometryFile": "../models/UnitBox.obj",
			"translation": [0,1,0],
			"rotationAxis": [1, 0, 0],
			"rotationAngle": 0,
			"scale": [2.6, 2.1, 1.0],
			"color": [0.1, 0.4, 0.6, 1.0], 
			"isDynamic": false,
			"isWall": true,
			"mapInvert": true, 
			"mapThickness": 0.0,
			"mapResolution": [30,25,20]
		}
	], 
	"Emitters": [
		{
			"width": 8, 
			"height": 3, 
			"translation": [-1.2,1,0.0],
			"rotationAxis": [1, 0, 0],
			"rotationAngle": 0,
			"velocity": 5,
			"type": 0,
			"emitEndTime": 2
		}
}

**Comment by janbender** [comment]
The width of your emitter is larger than the width of the box. You have 8 times the particle diameter (2*0.07) which is larger than the width of the box (1m). 

Note the volume maps push the particles away from the boundary so that I would recommend to use 16*0.07 + eps as box width.

**Reply by KimYoungGR** [reply]
Sorry the box size should be "scale": [3,3,2],

**Reply by janbender** [reply]
> Sorry the box size should be "scale": [3,3,2],

For this size the scene works fine for me.

**Reply by KimYoungGR** [reply]
In my project I need an precise fluid flow by using emitter. But  I tried some scenes, the total fluid volunm doesn't equal to total particle number * single particle volume, even the single particle volume was calculate as cubic.
I really hope you could explain this question.
Thank you.

**Reply by janbender** [reply]
Can you please go here: 

https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/master/SPlisHSPlasH/FluidModel.cpp#L258

and in line 260 and remove the factor of 0.8. We introduced this long ago to avoid pressure at the beginning of the simulation when you use a denser sampling. However, this also causes that the fluid with regular sampling has only 80% of the volume. 

We will remove this in future versions.

**Reply by KimYoungGR** [reply]
Thank you very much! I'll try it right away.

**Comment by janbender** [comment]
In fact since an SPH particle is used as sampling point for solving a volume integral, it is in this sense not a sphere but represents a cube with edge length = particle diameter. However, for rendering we naturally use sheres.
